### ISSUE TITLE
HTTP request or updating [(data)] will not keep array length > 1

### ISSUE THREAD

**Issue by lunchboxsushi** [hamzahamidi/ajsf#115]
**Describe the bug**
which template:

* [x] MaterialDesignFrameworkModule — Material Design
* [ ] Bootstrap3FrameworkModule — Bootstrap 3
* [ ] Bootstrap4FrameworkModule — Bootstrap 4
* [ ] NoFrameworkModule — plain HTML
* [ ] Other (please specify below)
  
* Array of string in schema will only render the first item in the list.

For examples HTTP request will return an data & schema.

Example
``` js
// schema
var schema = {
  type: "object",
  properties: {
    phone_number: {
      type: "array",
      items: {
        type: "string"
      }
    }
  }
}

var data = {
    phone_number: ["12321", 12312", "1321"]
}
```

However, when I hard code the object it behaves correctly, it's only when doing an HTTP request. that it will remove index 1,2 from data.phone_number. I can see it in Augury using $$el.componentInstance.data.phone_number it will no longer have the additional 2 items in the array..?

to do the HTTP request I use Angular HTTP Client with subscribe pattern nothing special, I even log it as soon as I as soon as I get the response in the success `(data) =>` call and it shows correctly the array has 3 items...

**To Reproduce**

Get the schema and data from a HTTP request where the schema & data is based on the example above.

**Expected behavior**
I expect add the data in the array to render and not only the first item.

**Screenshots**
snip

missing rendered items
![image](https://user-images.githubusercontent.com/10443153/59119092-e629c780-891f-11e9-8cde-0457ae5284ea.png)

log of HTTPrequest
![image](https://user-images.githubusercontent.com/10443153/59119137-06598680-8920-11e9-9c75-8b8f71aa969e.png)

Augury element right after it loads
![image](https://user-images.githubusercontent.com/10443153/59119195-238e5500-8920-11e9-94d9-6cda78c38527.png)

**Desktop (please complete the following information):**

- OS: [Windows 10]
- Browser [chrome@latest]

**Comment by lunchboxsushi** [comment]
After digging into the code base I found that after `setFormValues` is called once it hits `this.jsf.formGroup.patchValue(newFormValues);` it changes `this.jsf.data` to no longer contain any array items with more than 1 element... Not sure why in this case will keep looking

```js
JsonSchemaFormComponent.prototype.setFormValues = function (formValues, resetFirst) {
        if (resetFirst === void 0) { resetFirst = true; }
        if (formValues) {
            var newFormValues = this.objectWrap ? formValues['1'] : formValues;
            if (!this.jsf.formGroup) {
                this.jsf.formValues = formValues;
                this.activateForm();
            }
            else if (resetFirst) {
                this.jsf.formGroup.reset();
            }
            if (this.jsf.formGroup) {
                this.jsf.formGroup.patchValue(newFormValues);
            }
            if (this.onChange) {
                this.onChange(newFormValues);
            }
            if (this.onTouched) {
                this.onTouched(newFormValues);
            }
        }
        else {
            this.jsf.formGroup.reset();
        }
    };
```

**Comment by lunchboxsushi** [comment]
I believe it's due to a limitation of `angular.forms.PatchValue` [stackoverflow-patchValue](https://stackoverflow.com/a/40917579/5523033) currently the structure of the array wont change unless `push` is done and there's currently no `arrayPatchValue` any suggestions?

**Comment by lunchboxsushi** [comment]
As a work-around added a `*ngIf` around `json-schema-form` to only show after all data & schema is loaded from the HTTP request this way the `reset` is never called and `this.jsf.formGroup.reset()` is never called which will allow data integrity.

**Comment by samudiogo** [comment]
In my case, I made a custom select, then a I tried to resolve the HTTP request at own component, then a got a similar situation like yours. the select list item is only shows if a touch another input..

Now I will try your work-around! @lunchboxsushi , thanks for share it!

**Comment by samudiogo** [comment]
Hi @lunchboxsushi!
I am having problems to make it to works when I have multiple widget with dynamic schemas.. 
##### ngOnInit():

```typescript
 ngOnInit() {
  const sub =  this.route.paramMap.pipe(
      switchMap(params => this.service.geByCTIformItemId(params.get('form_id'), params.get('form_item_id'))),
      map(res => {
        this.resource = res;
        const row = {};

        if (this.route.snapshot.routeConfig.path !== ':form_id/novo') {
          const allowed = Object.keys(res.data).filter(f => !RegExp('^__').test(f));

          allowed.forEach(i => {
            row[i] = res.data[i];
          });
        }
        this.formData = row;

        this.formSchema = this.loadDependenciesFromUiSchema(res.json_schema, res.ui_schema);
        this.formSchema.formData = this.formData;
        return res;
      }),
    ).subscribe(res => {
       const validate = res.ui_schema.includes('source');
       !validate ? this.formActive = true : this.formActive = false;
      },
    );
  }
```
##### method to load other sources      
```typescript
private loadDependenciesFromUiSchema(json_schema: any, uiSchema: any): any {

        const jsonSchemaObj = JSON.parse(json_schema);
        const uiSchemaObj = JSON.parse(uiSchema);

        const uiItems = Object.keys(uiSchemaObj).filter(item => !item.includes(':'));

        uiItems.map(uiItem => {
          const widItem = uiSchemaObj[uiItem];

          if (widItem['ui:widget'] === 'select') {
            if (this.hasValidDynamic(widItem.options)) {
              const columnName = widItem.options.columnName;
              const columnValue = widItem.options.columnValue;
              this.http.get(widItem.options.source)
                .subscribe(res => {
                  this.formActive = false;
                  if (isArray(res)) {
                    const resArr = res as [];

                    widItem.options.titleMap = resArr.map(itemRes => {
                      return {
                        value: itemRes[columnValue],
                        name: itemRes[columnName],
                      };
                    });
                  } else if (isArray(res['data'])) {

                    widItem.options.titleMap = res['data'].map(itemRes => {
                      return {
                        value: itemRes[columnValue],
                        name: itemRes[columnName],
                      };
                    });
                  }
                  uiSchemaObj[uiItem] = widItem;

                  jsonSchemaObj.properties[uiItem].enum = widItem.options.titleMap.map(i => i.name);
                  this.formActive = true; // works only for unique component
                });
            } else {
              this.hasHttpRequest = false;
            }
          }
        });
        // this.formActive = true; //does not working if I put here
        return {
          schema: jsonSchemaObj,
          form: uiSchemaObj,
        };
      }
```

**Comment by lunchboxsushi** [comment]
When I did a deep dive into the source code, what I found is that if you update the data while the jsonSchema component is rendered (not in an *ngIf)  it will try to **update** vs **initialize**. When updating it will always remove the array items except the first one. This is main issue, my suggestion is to make sure you only have the jsonSchema in the dom rendered after all HTTP requests. If you try to update the *data* object in the future it will break the functionality.

At the very least that was my understanding perhaps I might be off on the underlying cause but verify that `JsonSchemaFormComponent.prototype.setFormValues = function (formValues, resetFirst)` is not being called, add a break point if it's hitting `this.jsf.formGroup.pathcValue(newFormValues)` then it will break functionality. 

Hope this helps.
